### PR TITLE
caveats optimization and stuff

### PR DESCRIPTION
**Pull request by thousandsofthem**
DO NOT MERGE
Another, much more efficient way of handling caveats.

Simple script for testing:
https://gist.github.com/thousandsofthem/b378e04cca1d9f7bf5f8

What's ok:
- serializing / unserializing structs in efficient manner, strings, []bytes, *ints, *uints, bools supported
- (un)marshaling to binary, json works with no issues

What's wrong:
- i'll add floats later if required
- `int`, `uint`s are not supported for safety - their size can vary from machine to machine
- tests are not up to date
- code is not cleaned up
- primary concern - verification failing for unknown reason. need help here :(

@edsrzf could  you take a look? something wrong with verification, don't know where i missed a screw

### CODE REVIEW

**Comment by edsrzf** [comment]
Can you summarize what you're doing here? It looks like two things:
- Now macaroons can only have a single caveat
- There's some struct serialization stuff added to the package, presumably for encoding caveats

I don't think limiting macaroons to one caveat is good because then we're getting rid of a big part of the appeal of macaroons, which is that caveats can be added after the macaroon is created.

I also don't think that serializing structs into caveats belongs in this package. It should be completely agnostic about what the caveat encoding format is, like the macaroon package it was forked from.

**Comment by edsrzf** [comment]
To clarify my thoughts on caveats, I'd like to see two changes:
- Compound caveats. That is, a single caveat that can represent multiple restrictions on the macaroon. If the package remains encoding-agnostic, this can be done outside the package.
- The ability to create a macaroon with a caveat. This way a macaroon with a caveat can be verified with only a single HMAC.

**Comment by thousandsofthem** [comment]
@edsrzf thanks, makes much more sense. i will 
1) slightly tune initial macaroon version to accept raw `[]byte` data instead of `string`
2) write, most likely, separate package for struct serialization
Done.
Sounds ok?

**Comment by edsrzf** [comment]
Yes, that sounds good. Thanks!
